feat(awscdk-construct): documentation

mbonig opened the pull request:
*Issue #, if available:*

*Description of changes:*

Added first draft doc for using the AWS CDK construct library. Notes requested.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,186 @@
+
+Use this if you'd like to build and distribute CDK constructs for others to use.
+
+# Getting Started
```
why not use `projen new`?

mbonig replied:
Yes, I'll add steps to use projen new instead of expecting this more manual approach.

eladb commented on docs/awscdk-construct.md:
````diff
@@ -0,0 +1,186 @@
+
+```typescript
+const project = new AwsCdkConstructLibrary({
+  authorAddress: "",
````
Instead of needing to maintain this, I rather we implement #168 and instruct people to use `npx projen new`

mbonig replied:
👍

eladb commented on docs/awscdk-construct.md:
````diff
@@ -0,0 +1,186 @@
+Next are getting CDK dependencies added:
+
+```typescript
+  entrypoint: 'lib/index.js',
````
How is `entrypoint` related to CDK dependencies? (also, I don't think it needs to be set up)

mbonig replied:
Totally forgot mentioning this. Since most people probably didn't name their construct's file 'main.ts' I figured this would need to be overridden. I'll make that clearer.

eladb commented on docs/awscdk-construct.md:
````diff
@@ -0,0 +1,186 @@
+  cdkTestDependencies: ["@aws-cdk/assert"],
+```
+
+With the above example you'll get a package.json that looks like this:
````
I don't think it's very useful to show the synthesized output for everything. Just describe the behavior. In this case, I would write something like:

`cdkDependencies` will add both `dependencies` and `peerDependencies` to your `package.json` file with a **caret** semver requirement. You can set `cdkVersionPinning` to `true` to use a fixed version, but this means that any consumer of your library will have to use this exact CDK version.

mbonig replied:
Will do. I was putting the output because that usually helps me understand how properties transfer but I can remove it.

eladb commented on docs/awscdk-construct.md:
````diff
@@ -0,0 +1,186 @@
+project.addCdkDependencies("aws-cdk/aws-sqs", "aws-cdk/aws-sns");
+```
+
+### JSII Packaging
````
This is not just packaging, it's also publishing.

mbonig replied:
But only if the related workflows are setup, right?

eladb replied:
No, workflows are automatically created as soon as you specify `python`, `java` etc. No additional setup is required (besides installing the secrets to your github repo).

eladb commented on docs/awscdk-construct.md:
````diff
@@ -0,0 +1,186 @@
+  },
+```
+
+Will render out the following to the package.json:
````
Again, redundant. It also modifies the GitHub workflows.

eladb commented on docs/awscdk-construct.md:
````diff
@@ -0,0 +1,186 @@
+```
+
+If you don't want to publish a particular package, do not include the `dotnet`, `java`, or `python` field.
+
````
In order to the github workflows to work, users need to add github secrets. This should be documented.

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,186 @@
+# AWS CDK - Construct library
```
```suggestion
# AWS CDK Construct Library
```

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,186 @@
+# AWS CDK - Construct library
+
+This projen type is used for building AWS CDK constructs using the JSII.
```
Explain what's the value of using jsii and include some references to jsii docs.

jsii is spelled all lowercase BTW

eladb reviewed (changes requested):
Initial round of feedback

mbonig commented:
I addressed all issues. I also added some additional coverage I didn't notice the first time around.

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,158 @@
+Review the resulting .projenrc.js file and make changes as needed. The following are some specific areas
+you may want to set explicitly.
+
+A `README.md` file is not created but encouraged. It is required for publishing your construct for Python (see [jsii publishing](#jsii-publishing)).
```
Should we bootstrap one?

mbonig replied:
I added a Readme class.

eladb commented on docs/awscdk-construct.md:
````diff
@@ -0,0 +1,158 @@
+Next are getting CDK dependencies added:
+
+```typescript
+  entrypoint: 'lib/our-s3-bucket.js',
````
Not sure what's up here? The practice that I'd recommend is leaving this as the default `lib/index.js` and adding this to `index.ts`:

```ts
export * from './our-s3-bucket';
```

mbonig replied:
👍

eladb commented on docs/awscdk-construct.md:
````diff
@@ -0,0 +1,158 @@
+  cdkTestDependencies: ["@aws-cdk/assert"],
+```
+
+The entrypoint can be set to point at your existing file, if not at `lib/index.js`. However, it's recommended that you
+rename your file to `index.js` and leave this un-set.
````
Remove

mbonig replied:
👍

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,158 @@
+requirement (e.g. `^1.67.0`). You can set cdkVersionPinning to true to use a fixed version, but this means that any 
+consumer of your library will have to use this exact CDK version. Likewise, `cdkTestDependencies` will add dependencies 
+to the `devDependencies`. 
+
```
Explain why both deps and peer deps are required (this is actually resolved by npm@7)

mbonig replied:
I assume changes are needed if projen is to support the new npm@7 algorithm and not add the cdk deps to `dependencies`?

eladb replied:
Yes, but it will take a while for users to migrate and this is a construct _library_, so we can't rely on people using npm@7.

Also, in CDK2.0, you'll only need a peer dependency.

mbonig replied:
k. I don't believe there is anything to add to this until we decide npm@7 is adopted enough.

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,158 @@
+
+### jsii Publishing
+
+As this is a [jsii](https://github.com/aws/jsii) constructs, it will cross-compile to other languages.  You can set up 
```
```suggestion
As this is a [jsii](https://github.com/aws/jsii) project, it will cross-compile to other languages.  You can set up 
```

mbonig replied:
👍

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,158 @@
+to be setup:
+
+* npm - NPM_TOKEN
+* .net - NUGET_API_KEY
```
```suggestion
* .NET - NUGET_API_KEY
```

mbonig replied:
👍

eladb commented on docs/awscdk-construct.md:
````diff
@@ -0,0 +1,158 @@
+```
+
+Each target requires a specific [Github project secret](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets) 
+to be setup:
````
Add a reference to [jsii-release](https://github.com/aws/jsii-release) which has detailed instructions on how to obtain these secrets.

mbonig replied:
👍

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,158 @@
+
+A `README.md` file is not created but encouraged. It is required for publishing your construct for Python (see [jsii publishing](#jsii-publishing)).
+
+### Standard Node Module Fields
```
Why H3 instead of H2?

mbonig replied:
missed some cleanup... fixed now.

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,158 @@
+This projen type is used for building AWS CDK constructs using the jsii. The [jsii](https://github.com/aws/jsii) allows 
+you to write code once in Typescript and it will generate the Python, .net, and Java equivalents. Use this if you'd 
+like to build and distribute CDK constructs for others to use.
+
```
Add TOC

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,158 @@
+* The Build workflow - controlled by the `buildWorkflow` field. On a 'pull_request' or 'workflow_dispatch' the construct
+will be built and checked for anti-tamper (ensure no manual changes to generated files).
+* The Release workflow - controlled by the `releaseWorkflow` field. On a push to `master` (overridden at
+ `props.defaultReleaseBranch`) the construct is built, anti-tampered, version bumped with a commit, pushed back to git, 
```
Replace the term "construct" across the doc with "construct library" or just "library"

mbonig replied:
I think I got the right ones. Some places still seemed 'construct' was best.

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,158 @@
+ `props.defaultReleaseBranch`) the construct is built, anti-tampered, version bumped with a commit, pushed back to git, 
+ and then published to the configured artifact repositories (e.g. npm, pypi)
+
+Additionally, a [dependabot](https://dependabot.com/) file will be added unless `dependabot` is set to 'false'.
```
This should be under "Dependencies", not "Workflows".

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,158 @@
+All are pretty standard setup and nothing CDK-specific at this point. The `keywords` automatically gets 'cdk' so you don't
+need to specify it. 
+
+### CDK Dependencies
```
```suggestion
### Dependencies
```

Please also add a section about non-CDK dependencies. Specifically, the difference between jsii deps and non-jsii deps (which require bundling)

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,158 @@
+These values are optional but allow the construct catalog's Twitter account to mention your handle on the tweet. Setting `announce`
+to 'false' will stop all tweets about the construct. However, the construct will still be indexed.
+
+### docgen
```
```suggestion
### API Documentation
```

mbonig replied:
👍

eladb reviewed (changes requested):
Missing:

- Enabling API compatibility checks
- Release management
- Project structure (src, lib, test, resources,...)
- Workflow (what is the meaning of each command).

Some of those should be part of the node/typescript/jsii project docs, but since we don't have them lets try to cover all of it here and then we can move stuff around.

Would be nice to start structuring our docs according to project types and then we can just cross reference.

mbonig commented:
> Missing:
> 
> * Enabling API compatibility checks

not sure what you're looking for here. More details about this? https://github.com/aws/jsii/tree/master/packages/jsii-diff

> * Release management

Added the Workflow section which should cover this, but let me know if there were other details you are thinking of.

> * Project structure (src, lib, test, resources,...)

Added. I didn't cover all the individual files, but added a comment about the resources directory.

> * Workflow (what is the meaning of each command).
> 

Added as `Scripts`

> Some of those should be part of the node/typescript/jsii project docs, but since we don't have them lets try to cover all of it here and then we can move stuff around.
> 
> Would be nice to start structuring our docs according to project types and then we can just cross reference.

I figured one thing that'll be useful is a breadcrumb at the top showing the hierarchy. Right now it doesn't link to any other docs but it should in the future (when more docs like this are built).

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,219 @@
+Review the resulting .projenrc.js file and make changes as needed. The following are some specific areas
+you may want to set explicitly.
+
+## Standard Node Module Fields
```
```suggestion
## Module Metadata
```

eladb commented on docs/awscdk-construct.md:
````diff
@@ -0,0 +1,219 @@
+These fields are your basic Node module setup:
+
+```typescript
+  authorAddress: "benisrae@amazon.com",
````
Please normalize all quotes to single quotes (I know that `projen new` generates with double...)

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,219 @@
+
+`cdkDependencies` will add both dependencies and peerDependencies to your package.json file with a caret semver 
+requirement (e.g. `^1.67.0`). CDK dependencies must be both direct and peer dependencies, 
+see [this issue](https://github.com/aws/aws-cdk/issues/5064). You can set cdkVersionPinning to true to use a fixed 
```
```suggestion
see [this issue](https://github.com/aws/aws-cdk/issues/5064). You can set `cdkVersionPinning` to `true` to use a fixed 
```

eladb commented on docs/awscdk-construct.md:
```diff
@@ -0,0 +1,219 @@
+
+## jsii Publishing
+
+As this is a [jsii](https://github.com/aws/jsii) project, it will cross-compile to other languages.  You can set up 
```
```suggestion
As this is a [jsii project](./jsii.md), it will cross-compile to other languages.  You can set up 
```

mbonig commented:
Anything else I missed?
